src/hydra-queue-runner/build-remote.cc

```diff
@@ -509,6 +555,23 @@ void State::buildRemote(ref<Store> destStore,
             result.overhead += std::chrono::duration_cast<std::chrono::milliseconds>(now2 - now1).count();
         }
 
+        /* Register the outputs of the newly built drv */
+        if (settings.isExperimentalFeatureEnabled(Xp::CaDerivations)) {
+          auto outputHashes = staticOutputHashes(*localStore, *step->drv);
+          for (auto & [outputId, realisation] : builtOutputs) {
+              // Register the resolved drv output
+              localStore->registerDrvOutput(realisation);
+              destStore->registerDrvOutput(realisation);
+
+              // Also register the unresolved one
+              auto unresolvedRealisation = realisation;
+              unresolvedRealisation.signatures.clear();
+              unresolvedRealisation.id.drvHash = outputHashes.at(outputId.outputName);
+              localStore->registerDrvOutput(unresolvedRealisation);
+              destStore->registerDrvOutput(unresolvedRealisation);
```

Comment:
```suggestion
              try {
                localStore->registerDrvOutput(realisation);
              } catch (Error &e) {
                printMsg(lvlWarn, "Failed registering realisation locally: %1% (already exists?)", e.what());
              }
              try {
                remoteStore->registerDrvOutput(realisation);
              } catch (Error &e) {
                printMsg(lvlWarn, "Failed registering realisation on ‘%1%’: %2% (already exists?)", machine->sshName, e.what());
              }

              // Also register the unresolved one
              auto unresolvedRealisation = realisation;
              unresolvedRealisation.signatures.clear();
              unresolvedRealisation.id.drvHash = outputHashes.at(outputId.outputName);
              try {
                localStore->registerDrvOutput(unresolvedRealisation);
              } catch (Error &e) {
                printMsg(lvlWarn, "Failed registering unresolved realisation locally: %1% (already exists?)", e.what());
              }
              try {
                remoteStore->registerDrvOutput(unresolvedRealisation);
              } catch (Error &e) {
                printMsg(lvlWarn, "Failed registering unresolved realisation on ‘%1%’: %2% (already exists?)", machine->sshName, e.what());
              }
```

Remote-building CA Derivations seems to fail here because the destStore already has the realisation registered. My hydra has this patch applied although you can probably remove the try {} catch for the localStore and pattern match the error to make sure it is about a constraint failure when inserting into Realisations

Reply:
This looks good to me, also removing the try/catch for the local store and pattern matching the other errors.
Do you already know how can we check that the error is caused by a constraint failure?

However I would prefer leaving the final word to @thufschmitt about this part.

Reply:
I don’t know enough about nix’s c++ code to know how to match a specific error

also it seems i was wrong about the “remove the localStore try catch blocks” because the error only seems to occur on the local parts my bad

```
Jul 02 10:02:54 nas hydra-queue-runner[1352608]: Failed registering realisation locally: error: executing SQLite statement '
Jul 02 10:02:54 nas hydra-queue-runner[1352608]:                        insert into Realisations (drvPath, outputName, outputPath, signatures)
Jul 02 10:02:54 nas hydra-queue-runner[1352608]:                        values ('sha256:4768517dfe0afebd04abc714af41ddd08785e3067abd98471ad2b608e8448a1f', 'out', (select id from ValidPaths where path = '/nix/store/cna07xr4yfh8ck0q00ibbpgqgmcmxrf4-system-units'), '')
Jul 02 10:02:54 nas hydra-queue-runner[1352608]:                        ;': constraint failed (in '/nix/var/nix/db/db.sqlite') (already exists?)
```

Reply:
Trying this out made me discover that the issue is not that the realisation already exists but that the out output is not copied over for some reason